File: scan-library-tool/perf.py
"""Performance logging — tracks per-phase timings during scan."""

import json
import os
import time
from datetime import datetime, timezone


class ScanTimer:
    def __init__(self, request_id: str, export_dir: str):
        self.request_id = request_id
        self.export_dir = export_dir
        self.start = time.monotonic()
        self.phases: list[dict] = []
        self._current: tuple[str, float] | None = None
        self._current_meta: dict | None = None

    def begin(self, name: str, meta: dict | None = None):
        now = time.monotonic()
        if self._current:
            self._end_current(now)
        self._current = (name, now)
        self._current_meta = meta

    def end(self):
        if self._current:
            self._end_current(time.monotonic())

    def _end_current(self, now: float):
        name, start = self._current
        elapsed_ms = round((now - start) * 1000)
        entry = {"phase": name, "elapsed_ms": elapsed_ms}
        if self._current_meta:
            entry["meta"] = self._current_meta
            self._current_meta = None
        self.phases.append(entry)
        self._current = None

    def finish(self) -> dict:
        self.end()
        total_ms = round((time.monotonic() - self.start) * 1000)
        result = {
            "request_id": self.request_id,
            "exported_at": datetime.now(tz=timezone.utc).isoformat(),
            "total_ms": total_ms,
            "phases": self.phases,
        }
        return result

    def write_log(self, folder: str, result: dict):
        path = os.path.join(folder, f"{self.request_id}_perf.json")
        with open(path, "w") as f:
            json.dump(result, f, indent=2)

File: scan-library-tool/rust/src/perf.rs
use std::time::Instant;
use serde::Serialize;
use chrono::Utc;

#[derive(Debug, Clone, Serialize)]
pub struct PhaseEntry {
    pub phase: String,
    pub elapsed_ms: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub meta: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize)]
pub struct ScanPerf {
    pub request_id: String,
    pub exported_at: String,
    pub total_ms: u64,
    pub phases: Vec<PhaseEntry>,
}

pub struct ScanTimer {
    request_id: String,
    start: Instant,
    phases: Vec<PhaseEntry>,
    current: Option<(String, Instant, Option<serde_json::Value>)>,
}

impl ScanTimer {
    pub fn new(request_id: String) -> Self {
        Self {
            request_id,
            start: Instant::now(),
            phases: Vec::new(),
            current: None,
        }
    }

    pub fn begin(&mut self, name: &str) {
        self.end_current();
        self.current = Some((name.to_string(), Instant::now(), None));
    }

    pub fn end(&mut self) {
        self.end_current();
    }

    fn end_current(&mut self) {
        if let Some((name, start, meta)) = self.current.take() {
            let elapsed_ms = start.elapsed().as_millis() as u64;
            self.phases.push(PhaseEntry {
                phase: name,
                elapsed_ms,
                meta,
            });
        }
    }

    pub fn record_phase(&mut self, name: &str, elapsed_ms: u64) {
        self.phases.push(PhaseEntry {
            phase: name.to_string(),
            elapsed_ms,
            meta: None,
        });
    }

    pub fn finish(&mut self) -> ScanPerf {
        self.end_current();
        let total_ms = self.start.elapsed().as_millis() as u64;
        ScanPerf {
            request_id: self.request_id.clone(),
            exported_at: Utc::now().to_rfc3339(),
            total_ms,
            phases: self.phases.clone(),
        }
    }
}
